File: lib/pelican_installer/screens/ssl.py
"""SSL certificate setup screen."""

from __future__ import annotations

from textual import on
from textual.app import ComposeResult
from textual.containers import Container, Vertical
from textual.screen import Screen
from textual.widgets import Button, Input, Static

from pelican_installer.utils.state import InstallState


class SSLScreen(Screen[str]):
    """Screen for SSL certificate configuration."""

    CSS = """
    SSLScreen {
        align: center middle;
    }

    #email-container {
        width: 100%;
        height: auto;
        margin-bottom: 1;
    }

    #email-input {
        width: 100%;
        margin-top: 1;
        margin-bottom: 1;
    }

    #ssl-info {
        color: #b7dbe8;
        margin-bottom: 1;
    }
    """

    def __init__(self, state: InstallState) -> None:
        super().__init__()
        self.state = state

    def compose(self) -> ComposeResult:
        with Container(id="root"):
            with Container(id="card"):
                yield Static("SSL Certificate Setup", id="title")
                yield Static(
                    "Let's Encrypt will generate free SSL certificates",
                    id="subtitle",
                )

                with Vertical(id="email-container"):
                    yield Static(
                        f"Domain: {self.state.domain}",
                        id="ssl-info",
                    )
                    yield Static("Email for certificate notifications:")
                    yield Input(
                        placeholder="admin@example.com",
                        id="email-input",
                    )

                yield Static(
                    "Enter email and press Next (n)",
                    id="hint",
                )

                with Container(id="footer"):
                    yield Button("Back (b)", id="back")
                    yield Button("Close (c)", id="close")
                    yield Button("Next (n)", id="next", disabled=True)

    def on_mount(self) -> None:
        """Focus the input on mount."""
        self.query_one("#email-input", Input).focus()

    @on(Input.Changed, "#email-input")
    def validate_input(self, event: Input.Changed) -> None:
        """Enable Next button when email is valid."""
        next_btn = self.query_one("#next", Button)
        # Simple email validation
        value = event.value.strip()
        next_btn.disabled = "@" not in value or "." not in value

    @on(Input.Submitted, "#email-input")
    def input_submitted(self) -> None:
        """Handle Enter key in input."""
        self._proceed_next()

    @on(Button.Pressed, "#next")
    def next_pressed(self) -> None:
        """Handle Next button."""
        self._proceed_next()

    def _proceed_next(self) -> None:
        """Validate and proceed to dependencies."""
        email_input = self.query_one("#email-input", Input)
        email = email_input.value.strip()

        if "@" in email and "." in email:
            self.state.ssl_email = email
            self.dismiss("dependencies")

    @on(Button.Pressed, "#back")
    def back_pressed(self) -> None:
        """Go back to domain configuration."""
        self.dismiss("back")

    @on(Button.Pressed, "#close")
    def close_pressed(self) -> None:
        """Handle close button."""
        self.app.exit()

    def action_request_close(self) -> None:
        """Global close action (c key)."""
        self.app.exit()

File: src/screens/ssl.rs
use crate::screens::{Screen, ScreenResult};
use crate::ui::components::TextInput;
use crate::utils::state::InstallState;
use ratatui::prelude::*;
use ratatui::widgets::Paragraph;
use crossterm::event::KeyEvent;

pub struct SSLScreen {
    input: TextInput,
    can_proceed: bool,
}

impl SSLScreen {
    pub fn new() -> Self {
        Self {
            input: TextInput::new("admin@example.com"),
            can_proceed: false,
        }
    }
}

impl Screen for SSLScreen {
    fn on_mount(&mut self, _state: &mut InstallState) -> crate::error::Result<()> {
        self.input.set_focused(true);
        Ok(())
    }
    
    fn render(&mut self, area: Rect, frame: &mut Frame, state: &InstallState) -> crate::error::Result<()> {
        let chunks = Layout::default()
            .direction(Direction::Vertical)
            .margin(1)
            .constraints([
                Constraint::Length(3),  // Title
                Constraint::Length(2),  // Subtitle
                Constraint::Length(2),  // Domain info
                Constraint::Length(5),  // Input
                Constraint::Length(2),  // Hint
                Constraint::Length(3),  // Footer
            ])
            .split(area);
        
        // Title
        let title = Paragraph::new("SSL Certificate Setup")
            .style(Style::default().fg(Color::Cyan).add_modifier(Modifier::BOLD))
            .alignment(Alignment::Center);
        frame.render_widget(title, chunks[0]);
        
        // Subtitle
        let subtitle = Paragraph::new("Let's Encrypt will generate free SSL certificates")
            .style(Style::default().fg(Color::Gray))
            .alignment(Alignment::Center);
        frame.render_widget(subtitle, chunks[1]);
        
        // Domain info
        let domain_info = Paragraph::new(format!("Domain: {}", state.domain))
            .style(Style::default().fg(Color::Cyan))
            .alignment(Alignment::Center);
        frame.render_widget(domain_info, chunks[2]);
        
        // Input label
        let input_label = Paragraph::new("Email for certificate notifications:")
            .style(Style::default().fg(Color::Gray))
            .alignment(Alignment::Center);
        frame.render_widget(input_label, Rect {
            x: chunks[3].x,
            y: chunks[3].y,
            width: chunks[3].width,
            height: 1,
        });
        
        // Input
        self.input.render(Rect {
            x: chunks[3].x,
            y: chunks[3].y + 1,
            width: chunks[3].width,
            height: 3,
        }, frame.buffer_mut());
        
        // Hint
        let hint = Paragraph::new("Enter email and press Next (n)")
            .style(Style::default().fg(Color::Gray))
            .alignment(Alignment::Center);
        frame.render_widget(hint, chunks[4]);
        
        // Footer
        let footer = Paragraph::new("Back: b | Next: n | Close: c")
            .style(Style::default().fg(Color::Yellow))
            .alignment(Alignment::Center);
        frame.render_widget(footer, chunks[5]);
        
        Ok(())
    }
    
    fn handle_key(&mut self, key: KeyEvent, state: &mut InstallState) -> crate::error::Result<ScreenResult> {
        use crossterm::event::KeyCode;
        
        match key.code {
            KeyCode::Char(c) => {
                self.input.handle_char(c);
                self.validate();
            }
            KeyCode::Backspace => {
                self.input.backspace();
                self.validate();
            }
            KeyCode::Delete => {
                self.input.delete();
                self.validate();
            }
            KeyCode::Left => self.input.move_left(),
            KeyCode::Right => self.input.move_right(),
            KeyCode::Home => self.input.move_home(),
            KeyCode::End => self.input.move_end(),
            KeyCode::Enter | KeyCode::Char('n') => {
                if self.can_proceed {
                    state.ssl_email = self.input.value().to_string();
                    return Ok(ScreenResult::Next(Box::new(super::install::InstallScreen::new())));
                }
            }
            KeyCode::Char('b') => return Ok(ScreenResult::Back),
            KeyCode::Char('c') => return Ok(ScreenResult::Exit),
            _ => {}
        }
        Ok(ScreenResult::None)
    }
}

impl SSLScreen {
    fn validate(&mut self) {
        let value = self.input.value().trim();
        self.can_proceed = value.contains('@') && value.contains('.');
    }
}
